### ISSUE TITLE
Fix/Replace/Upgrade Python3 dependencies

### ISSUE THREAD

**Issue by eviljeff** [mozilla/addons-server#4997]
https://pyup.io/account/repos/github/mozilla/addons-server/
Old dependencies that aren't compatible are blocking our smooth and totally uncomplicated upgrade of addons-server codebase to Python3, so we need to fix those first.

Work needed for prod / local development:
* [x] schematic ==0.4 - mozilla/schematic#4 (but obsolete after django migrations #3705 anyway)

Obsolete so will be removed:
* [x] amo-validator ==1.10.55 mozilla/addons-server#10502
  * [x] fastchardet ==0.2.0
  * [x] spidermonkey ==54.0a2.post2 mozilla/python-spidermonkey/pull/9

False positives:
* [x] filelock ==3.0.10 - travis.yml indicates it is py3 compatible.
* [x] futures ==3.2.0 - just a backport of py3 functionality to py2
* [x] pathtools ==0.1.2 - pretty unmaintained but is seemingly only used by Celery, which _is_ py3 compatible so safe to ignore.
* [x] hawkauthlib ==2.0.0  - repo pull requests and tox.ini indicate it is py3 compatible.
* [x] PyFxA ==0.5.0  - repo pull requests and travis.yml indicate it is py3 compatible.

Finished, changes merged into addon-server:
* [x] ~atomicwrites ==1.1.5~
* [x] ~basket-client ==0.3.12 - _might_ be so simple it's py3 compatible anyway.~
* [x] ~django-cache-nuggets ==0.1.1 #4998~
* [x] ~django-cronjobs ==0.2.3 #5050~
* [x] ~django-mobility (https://github.com/jbalogh/django-mobility) - only used for mobile pages #5055~
* [x] ~django-quieter-formset ==0.4 #5005~
* [x] ~django-multidb-router ==0.6 https://github.com/jbalogh/django-multidb-router/pull/31~
* [x] ~django-statsd-mozilla ==0.3.16 https://github.com/django-statsd/django-statsd/pull/98~
* [x] ~flake8-mutable ==1.1.0 https://github.com/ebeweber/flake8-mutable/pull/10~
* ~flufl.lock 2.4.1 - only supports python 2 and version 3.2 only supports python 3. Separate py3 requirements file added.~
* [x] ~fxapom ==1.10.0 https://github.com/mozilla/fxapom/issues/97~
* [x] ~jingo-minify ==0.6.0 folded into addons-server~
* [x] ~moz-addon-packager ==1.0.1 #5012~
* [x] ~MySQL-python ==1.2.5 #5002~
* [x] ~M2Crypto ==0.24.0 https://github.com/mozilla/addons-server/pull/5801~
  * [x] ~django-aesfield ==0.3.0 https://github.com/andymckay/django-aesfield/pull/10~
  * [x] ~m2secret ==0.1.1 replaced with https://pypi.python.org/pypi/m2secret-py3~
  * [x] ~signing-clients ==0.1.14 https://github.com/mozilla/signing-clients/pull/19~
* [x] ~mozilla-logger ==0.2 - Got removed once we switched completely to Logging 2.0 (See #4931)~
* [x] ~sphinxcontrib-httpdomain ==1.5.0~
* [x] ~ supervisor ==3.3.4 - master branch is py3 compatible so switched to that as it's only used in local development. #10339 ~

**Comment by EnTeQuAk** [comment]
Plan to get rid of M2crypto:

* [x] Port m2secret to cryptography to support python 3
 * done: https://pypi.python.org/pypi/m2secret-py3
* [ ] signing-clients: https://github.com/mozilla/signing-clients/pull/19
* [ ] port django-aesfield to use new m2secret + to python 3 (and more recent django apis)

**Comment by EnTeQuAk** [comment]
I started working on https://github.com/mozilla/addons-server/tree/1573-implement-django-jinja to implement django-jinja and replace jingo and eventually replace jingo-minify in our code, might bear more fruits next week or on my flight to SFO.

Otherwise, new django-aesfield is incoming and we started requesting proper permissions to release signing_clients so this should be incoming soon too I hope.

**Comment by eviljeff** [comment]
New python3 blockers since last time I looked at this post 😞 

supervisor ==3.3.4
pathtools ==0.1.2

hawkauthlib ==2.0.0
PyFxA ==0.5.0

**Comment by eviljeff** [comment]
One new extra python3 blocker:

basket-client ==0.3.12

And one false blocker (because it's just a backport of py3 functionality):

futures ==3.2.0

**Comment by EnTeQuAk** [comment]
Added a note about flufl.lock

**Comment by eviljeff** [comment]
Updated the description.  Blockers remaining (that will require some work from us) at this point are:
* basket-client
* schematic

**Comment by eviljeff** [comment]
And for the UI tests (they don't strictly block py3 support for prod -though would probably block dropping py2 support in code):
* hawkauthlib
* PyFxA

**Comment by eviljeff** [comment]
From a look through repo for hawkauthlib, it looks like it _is_ python3 compatible now, it just doesn't' indicate as such in setup.py.  PyFxA also.  So no UI test work needed (though it'd be cool if the setup files for those repos would indicate python version compatibility.

**Comment by EnTeQuAk** [comment]
> though it'd be cool if the setup files for those repos would indicate python version compatibility.

Sounds like a proper Friday task.

**Comment by diox** [comment]
`supervisor` is only required by `dev.txt`, maybe we could get away with installing it from a known commit directly, since there are no prerelease packages yet?

**Comment by EnTeQuAk** [comment]
> supervisor is only required by dev.txt, maybe we could get away with installing it from a known commit directly, since there are no prerelease packages yet?

Do you mean a commit to a supervisor source-code that is python3 compatible?

I'd maybe use circus instead of supervisor which works mostly similar for our use-case.

**Comment by diox** [comment]
In the meantime, I made a PR for basket-client: https://github.com/mozilla/basket-client/pull/33

**Comment by eviljeff** [comment]
Once mozilla/schematic#4 is merged (and we've got the dependency update) and #10502 is done we can (finally) call this closed as the remaining packages are python3 compatible.

**Comment by EnTeQuAk** [comment]
amo-validator is gone now!

and schematic is in addons-server now!

Which means… we're done? Need to check pyup to make sure…

**Comment by diox** [comment]
As a final task, maybe we could raise pull requests against the projects that don't advertise their Python 3 support ? Adding the relevant info to the `classifiers` in their `setup.py` files ?

**Comment by EnTeQuAk** [comment]
> As a final task, maybe we could raise pull requests against the projects that don't advertise their Python 3 support ? Adding the relevant info to the classifiers in their setup.py files ?

We should definitely do that.

**Comment by eviljeff** [comment]
I'll file a follow-up for raising prs.  Closing this issue as all the work is done.  Great work us!
